src/hocon_cli.erl

```diff
@@ -18,39 +18,265 @@
+        {error, E} ->
+            ?LOG:error("~p~n", [E]),
+            stop_deactivate();
+        {ok, Conf} ->
+            Conf
     end.
 
-to_map(Opts) ->
-    to_map(Opts, #{}).
-
-to_map([], Acc) -> Acc;
-to_map([A | T], Acc) when is_atom(A) ->
-    to_map(T, Acc#{A => true});
-to_map([{K, V} | T], Acc) ->
-    to_map(T, Acc#{K => V}).
-
-do(#{help := _}) ->
-    getopt:usage(spec(), "hocon", "");
-do(#{file := File}) ->
-    case hocon:load(File) of
-        {ok, Config} -> ok = pp_eterm(Config);
-        {error, Reason} -> exit(Reason)
+-spec writable_destination_path([proplists:property()]) -> file:filename() | error.
+writable_destination_path(ParsedArgs) ->
+    EtcDir = proplists:get_value(etc_dir, ParsedArgs),
+    DestinationPath = proplists:get_value(dest_dir, ParsedArgs, filename:join(EtcDir, "generated")),
+    AbsoluteDestPath = case DestinationPath of
+                           [$/ | _] -> DestinationPath;
+                           _      -> filename:join(element(2, file:get_cwd()), DestinationPath)
+                       end,
+    %% Check Permissions
+    case filelib:ensure_dir(filename:join(AbsoluteDestPath, "weaksauce.dummy")) of
+        %% filelib:ensure_dir/1 requires a dummy filename in the argument,
+        %% I think that is weaksauce, hence "weaksauce.dummy"
+        ok ->
+            AbsoluteDestPath;
+        {error, E} ->
+            ?LOG:error(
+                "Error creating ~s: ~s",
+                [AbsoluteDestPath, file:format_error(E)]),
+            error
     end.
 
-print(IoData) -> io:put_chars(IoData).
+-spec engage_hocon([proplists:property()]) -> {string(), string()} | error.
+engage_hocon(ParsedArgs) ->
+    AbsPath = case writable_destination_path(ParsedArgs) of
+                  error ->
+                      stop_deactivate(),
+                      error;
+                  Path -> Path
+              end,
+
+    Date = calendar:local_time(),
+
+    DestFile = proplists:get_value(dest_file, ParsedArgs),
+
+    DestinationFilename = filename_maker(DestFile, Date, "config"),
+    Destination = filename:join(AbsPath, DestinationFilename),
+
+    DestinationVMArgsFilename = filename_maker("vm", Date, "args"),
+    DestinationVMArgs = filename:join(AbsPath, DestinationVMArgsFilename),
+    ?LOG:debug("Generating config in: ~p", [Destination]),
+
+    Schema = load_schema(ParsedArgs),
+    Conf = load_conf(ParsedArgs),
+    LogFun = case proplists:get_value(verbose_env, ParsedArgs) of
+                 true ->
+                     fun(Key, Value) -> ?STDOUT("~s = ~p", [string:join(Key, "."), Value]) end;
+                 false ->
+                     fun(_, _) -> ok end
+             end,
+
+    case hocon_schema:generate(Schema, Conf) of
+        %{error, _X} ->
+            % @TODO print error
+        %    error;
+        NewConfig ->
+            AppConfig = proplists:delete(vm_args, NewConfig),
+            VmArgs = stringify(proplists:get_value(vm_args, NewConfig)),
```

Comment:
Since the data is going to be written into a file, you can avoid `lists:flatten` and `string:join` calls here. `file:write_file` function accepts iolists as inputs, and it's a better data structure performance-wise, because you'll avoid extra list concatentations.